File: rust/src/leetcode/Q208_implement_trie_prefix_tree.rs
/*
 * @lc app=leetcode id=208 lang=rust
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<char, Box<TrieNode>>,
    word: bool,
}

impl TrieNode {
    #[inline]
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            word: false,
        }
    }
}

struct Trie {
    start: TrieNode,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {
    fn new() -> Self {
        Trie {
            start: TrieNode::new(),
        }
    }

    fn insert(&mut self, word: String) {
        let mut current: &mut TrieNode = &mut self.start;

        for c in word.chars() {
            current = current
                .children
                .entry(c)
                .or_insert(Box::new(TrieNode::new()));
        }

        current.word = true;
    }

    fn search(&self, word: String) -> bool {
        let mut current: &TrieNode = &self.start;

        for c in word.chars() {
            if let Some(next_node) = current.children.get(&c) {
                current = next_node;
            } else {
                return false;
            }
        }

        current.word
    }

    fn starts_with(&self, prefix: String) -> bool {
        let mut current: &TrieNode = &self.start;

        for c in prefix.chars() {
            if let Some(next_node) = current.children.get(&c) {
                current = next_node;
            } else {
                return false;
            }
        }

        true
    }
}

// /**
//  * Your Trie object will be instantiated and called as such:
//  * let obj = Trie::new();
//  * obj.insert(word);
//  * let ret_2: bool = obj.search(word);
//  * let ret_3: bool = obj.starts_with(prefix);
//  */
// @lc code=end

File: python/Q208_implement_trie_prefix_tree.py
#
# @lc app=leetcode id=208 lang=python3
#
# [208] Implement Trie (Prefix Tree)
#


# @lc code=start
class TrieNode:
    def __init__(self) -> None:
        self.children: dict[str, TrieNode] = dict()
        self.end: bool = False


class Trie:
    def __init__(self) -> None:
        self.words: TrieNode = TrieNode()

    def insert(self, word: str) -> None:
        current: TrieNode = self.words

        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()

            current = current.children[c]

        current.end = True

    def search(self, word: str) -> bool:
        current: TrieNode = self.words

        for c in word:
            if c not in current.children:
                return False

            current = current.children[c]

        return current.end

    def startsWith(self, prefix: str) -> bool:
        current: TrieNode = self.words

        for c in prefix:
            if c not in current.children:
                return False

            current = current.children[c]

        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end
